File: src/main/java/leetcode/dynamic/n_queen.py
# -*- coding: utf-8 -*-


class NQueen:
    result = []
    cols = set()
    pie = set()
    na = set()

    def queen(self, n):
        if n < 1:
            return []

        self.dfs(n, 0, [])

        return self.result

    def dfs(self, n, row, cur_state):
        if row >= n:
            self.result.append(cur_state)
            return

        for col in range(n):
            if col in self.cols or row + col in self.pie or col-row in self.na:
                continue

            self.cols.add(col)
            self.pie.add(row+col)
            self.na.add(col-row)

            self.dfs(n, row+1, cur_state+[col])

            self.cols.remove(col)
            self.pie.remove(row + col)
            self.na.remove(col-row)

    def print_result(self, n):
        for res in results:
            for i in res:
                for j in range(n):
                    if i == j:
                        print("Q", end=" ")
                    else:
                        print("*", end=" ")
                print()
            print()


if __name__ == '__main__':
    q = NQueen()
    results = q.queen(4)
    q.print_result(4)

    l1 = []
    l2 = [3]
    print(l1.append(l2))

File: src/main/java/problems/backtrack/NQueen.java
package problems.backtrack;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class NQueen {
    //int[] result = new int[4];
    private List<List<Integer>> result = new ArrayList<>();
    private Set<Integer> cols = new HashSet<>();
    private Set<Integer> pie = new HashSet<>();
    private Set<Integer> na = new HashSet<>();

    public static void main(String[] args) {
        NQueen nQueen = new NQueen();
        System.out.println(nQueen.solveNQueens(4));
    }

    public List<List<Integer>> solveNQueens(int n) {
        if (n < 1)
            return new ArrayList<>();

        dfs(n, 0, new ArrayList<>());

        return result;
    }

    public void dfs(int n, int row, ArrayList<Integer> curState) {
        if (row >= n) {
            result.add(curState);
            return;
        }

        for (int col = 0; col < n; col++) {
            if (cols.contains(col) || pie.contains(row + col) || na.contains(col - row))
                continue;

            cols.add(col);
            pie.add(row + col);
            na.add(col - row);

            curState.add(col);
            dfs(n, row+1, curState);

            cols.remove(col);
            pie.remove(row + col);
            na.remove(col - row);
        }
    }

//    public void printResult (int n) {
//        List<String> board = new ArrayList<>();
//        for (int i = 0; i < result.size(); i++) {
//            for (int j = 0; j < result.get(i).size(); j++) {
//                board.add()
//
//            }
//        }
//    }




//    //逐行扫描
//    public void solveNQueens(int row) {
//        if (row == 4) {
//            printQueens(result);
//            return;
//        }
//
//        for (int column = 0; column < 4; ++column) {
//            if (isOk(row, column)) {
//                result[row] = column;
//                solveNQueens(row + 1);
//            }
//        }
//    }
//
//    private boolean isOk(int row, int column) {
//        int leftup = column - 1;
//        int rightup = column + 1;
//
//        for (int i = row - 1; i >= 0; i--) {
//            if (result[i] == column) return false;
//
//            if (leftup >= 0 && result[i] == leftup) return false;
//
//            if (rightup < 4 && result[i] == rightup) return false;
//
//            --leftup;
//            ++rightup;
//        }
//        return true;
//    }
//
//    private void printQueens(int[] result) {
//        for (int row = 0; row < 4; ++row) {
//            for (int column = 0; column < 4; ++column) {
//                if (result[row] == column) {
//                    System.out.print("O ");
//                } else {
//                    System.out.print("* ");
//                }
//            }
//            System.out.println();
//        }
//        System.out.println();
//    }
}
